File: packages/shared/src/knowledge-inbox.ts
import type { ContextPackEvidenceRef, EvolutionClaim, EvolutionCrossReview, InsightRevision } from "./evolution/index";
import type { ExperienceRecord, ExperienceFeedback } from "./experience";

/** Consumption only: these operations never authorize or invalidate knowledge. */
export type InboxSource = "evolution" | "experience";
export type InboxState = "pending" | "processed";
export interface InboxListQuery {
  state: InboxState;
  /** Omit to include both sources. Filtering happens before pagination. */
  source?: InboxSource;
  repositoryId?: string;
  limit?: number;
  cursor?: string;
}
export interface InboxRepository {
  repositoryId: string;
  displayName: string;
  projectAliases: string[];
}
export interface InboxContent {
  statement: string;
  applicability: string;
  guidance?: string[];
  actions?: string[];
  avoid?: string[];
  verification?: string[];
}
export interface InboxItem extends InboxContent {
  itemId: string;
  repositoryId: string;
  projectName: string;
  source: InboxSource;
  sourceId: string;
  sourceRevision: string;
  /** Exact source revision, independent of the public body's version hash. */
  sourceRevisionId?: string;
  kind: "finding" | "suggestion" | "experience";
  title: string;
  validationLabel: string;
  contentVersion: string;
  contentUpdatedAt: string;
  stateVersion: number;
  processedAt: string | null;
  hasUpdate: boolean;
  availability: "available" | "unavailable" | "unknown";
  /** Present on a processed snapshot when the current content has changed. */
  currentContentVersion?: string;
}
export interface InboxSourceStatus {
  status: "ok" | "partial";
  evolution: "available" | "unavailable";
  experience: "available" | "unavailable";
}
export interface InboxPage {
  items: InboxItem[];
  nextCursor: string | null;
  sourceStatus: InboxSourceStatus;
  repositories: InboxRepository[];
}
export interface InboxStateChange {
  state: InboxState;
  expectedContentVersion: string;
  expectedStateVersion: number;
}

/** An authenticated locator, never a bearer credential or an arbitrary URL. */
export const KNOWLEDGE_REFERENCE_PATTERN = /^rw-knowledge:v1:([a-f0-9-]{36}):([a-f0-9-]{36})$/u;
export interface KnowledgeShareRequest {
  contentVersion: string;
  sourceRevision: string;
}
export interface KnowledgeShareResult {
  reference: string;
  text: string;
}
export interface KnowledgeMaterial {
  warnings: string[];
  evolution?: {
    revision: InsightRevision;
    claims: EvolutionClaim[];
    reviews: EvolutionCrossReview[];
    evidence: ContextPackEvidenceRef[];
  };
  experience?: {
    record: ExperienceRecord;
    feedback: ExperienceFeedback[];
  };
}
export interface KnowledgeSnapshot {
  createdAt: string;
  item: InboxItem;
  material: KnowledgeMaterial;
}
export interface KnowledgeReadResult {
  reference: string;
  createdAt: string;
  item: InboxItem;
  current: {
    availability: InboxItem["availability"];
    contentVersion?: string;
    sourceRevision?: string;
    reason?: string;
  };
  material?: KnowledgeMaterial;
  /** Live Activity contents, fetched with normal authenticated content reads by the CLI. */
  contents?: Array<{
    contentId: string;
    status: "available" | "unavailable";
    text?: string;
    truncated?: boolean;
    reason?: string;
  }>;
}

File: packages/app-ios/Sources/RunweaveIOS/Contracts/KnowledgeInbox.swift
import Foundation

struct KnowledgeShareResult: Decodable {
  let reference: String
  let text: String
}

struct InboxRepository: Decodable, Identifiable {
  let repositoryId: String
  let displayName: String
  let projectAliases: [String]
  var id: String { repositoryId }
}
struct InboxItem: Decodable, Identifiable {
  let itemId: String
  let repositoryId: String
  let projectName: String
  let source: String
  let sourceId: String
  let sourceRevision: String
  let kind: String
  let title: String
  let statement: String
  let applicability: String
  let guidance: [String]?
  let actions: [String]?
  let avoid: [String]?
  let verification: [String]?
  let validationLabel: String
  let contentVersion: String
  let contentUpdatedAt: String
  let stateVersion: Int
  let processedAt: String?
  let hasUpdate: Bool
  let availability: String
  let currentContentVersion: String?
  var id: String { itemId }
  var excerpt: String { statement.isEmpty ? (actions?.first ?? applicability) : statement }
  var sourceLabel: String { source == "experience" ? "经验" : kind == "suggestion" ? "建议" : "洞察" }
  var updatedLabel: String {
    let formatter = ISO8601DateFormatter()
    formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
    guard let date = formatter.date(from: contentUpdatedAt) else { return contentUpdatedAt }
    return date.formatted(date: .abbreviated, time: .shortened)
  }
}
struct InboxPage: Decodable {
  let items: [InboxItem]
  let nextCursor: String?
  let sourceStatus: InboxSourceStatus
  let repositories: [InboxRepository]
}
struct InboxSourceStatus: Decodable {
  let status: String
  let evolution: String
  let experience: String
}
